core/plugins.rs

```diff
@@ -0,0 +1,82 @@
+use crate::isolate::Op;
+use crate::libdeno::PinnedBuf;
+use std::fmt;
+use std::io;
+
+pub type PluginResult<T> = std::result::Result<T, PluginError>;
+
+#[derive(Debug)]
+pub struct PluginError {
+  repr: Repr,
+}
+
+#[derive(Debug)]
+enum Repr {
+  Simple(String),
+  IoErr(io::Error),
+}
+
+pub fn new_plugin_error(msg: String) -> PluginError {
+  PluginError {
+    repr: Repr::Simple(msg),
+  }
+}
+
+impl fmt::Display for PluginError {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    match self.repr {
+      Repr::Simple(ref err_str) => f.pad(err_str),
+      Repr::IoErr(ref err) => err.fmt(f),
+    }
+  }
+}
+
+impl std::error::Error for PluginError {
+  fn description(&self) -> &str {
+    match self.repr {
+      Repr::Simple(ref msg) => msg.as_str(),
+      Repr::IoErr(ref err) => err.description(),
+    }
+  }
+
+  fn cause(&self) -> Option<&dyn std::error::Error> {
+    match self.repr {
+      Repr::Simple(ref _msg) => None,
+      Repr::IoErr(ref err) => Some(err),
+    }
+  }
+}
+
+impl From<io::Error> for PluginError {
+  #[inline]
+  fn from(err: io::Error) -> Self {
+    Self {
+      repr: Repr::IoErr(err),
+    }
+  }
+}
+
+/// Base result type for a plugin op represents either a Sync or Async value
+pub type PluginOp = Op<PluginError>;
+
+/// Complete return type for a plugin op including Sync errors
+pub type PluginOpResult = PluginResult<PluginOp>;
+
+/// Funciton type for plugin ops
+pub type PluginDispatchFn =
+  fn(is_sync: bool, data: &[u8], zero_copy: Option<PinnedBuf>)
+    -> PluginOpResult;
```

Comment:
I think this should be 
```rust
pub type PluginDispatchFn =
  fn(control_data: &[u8], zero_copy: Option<PinnedBuf>) -> Op<()>;
```

Reply:
Errors?

Reply:
I think errors should be handled by the plugin itself.

I see you're concerned about the ergonomics of implementing the op ... having access to `?` operator and whatnot. But those things can be shoehorned into this model with some utilities.

Reply:
Also how do you want to handle sync/async?

Reply:
The user should encode this into the buffers somehow.

Reply:
How does `sendInternal` know what to expect?

Reply:
`sendInternal()` is a private function called from `sendAsync()` or `sendSync()` 

Only some ops are both sync and async. For example `Deno.isTTY()` is strictly sync
https://github.com/denoland/deno/blob/c56df45355c8e68eabbfa62021e7ca7484115c0b/js/os.ts#L30-L33